arch/sim/src/sim/up_uart.c

```diff
@@ -0,0 +1,570 @@
+
+static int tty_setup(FAR struct uart_dev_s *dev)
+{
+  FAR struct tty_priv_s *priv = dev->priv;
+
+  if (!dev->isconsole && priv->fd < 0)
+    {
+      priv->fd = simuart_open(priv->path);
+      if (priv->fd < 0)
+        {
+          return priv->fd;
+        }
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: tty_shutdown
+ *
+ * Description:
+ *   Disable the UART.  This method is called when the serial
+ *   port is closed
+ *
+ ****************************************************************************/
+
+static void tty_shutdown(struct uart_dev_s *dev)
+{
+  FAR struct tty_priv_s *priv = dev->priv;
+
+  if (!dev->isconsole && priv->fd >= 0)
+    {
+      /* close file Description and reset fd */
+
+      simuart_close(priv->fd);
+      priv->fd = -1;
+    }
+}
+
+/****************************************************************************
+ * Name: tty_attach
+ *
+ * Description:
+ *   Configure the UART to operation in interrupt driven mode.  This method
+ *   is called when the serial port is opened.  Normally, this is just after
+ *   the setup() method is called, however, the serial console may operate in
+ *   a non-interrupt driven mode during the boot phase.
+ *
+ *   RX and TX interrupts are not enabled when by the attach method (unless
+ *   the hardware supports multiple levels of interrupt enabling).  The RX
+ *   and TX interrupts are not enabled until the txint() and rxint() methods
+ *   are called.
+ *
+ ****************************************************************************/
+
+static int tty_attach(struct uart_dev_s *dev)
+{
+  return OK;
+}
+
+/****************************************************************************
+ * Name: tty_detach
+ *
+ * Description:
+ *   Detach UART interrupts.  This method is called when the serial port is
+ *   closed normally just before the shutdown method is called.  The
+ *   exception is the serial console which is never shutdown.
+ *
+ ****************************************************************************/
+
+static void tty_detach(FAR struct uart_dev_s *dev)
+{
+}
+
+/****************************************************************************
+ * Name: tty_ioctl
+ *
+ * Description:
+ *   All ioctl calls will be routed through this method
+ *
+ ****************************************************************************/
+
+static int tty_ioctl(struct file *filep, int cmd, unsigned long arg)
+{
+#ifdef CONFIG_SERIAL_TERMIOS
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct uart_dev_s *dev = inode->i_private;
+  FAR struct tty_priv_s *priv = dev->priv;
+  FAR struct termios *termiosp = (FAR struct termios *)(uintptr_t)arg;
+
+  if (!termiosp)
+    {
+      return -EINVAL;
+    }
+
+  switch (cmd)
+    {
+      case TCGETS:
+        return simuart_getcflag(priv->fd, &termiosp->c_cflag);
+
+      case TCSETS:
+        return simuart_setcflag(priv->fd, termiosp->c_cflag);
```

Comment:
should we check dev->isconsole?

Reply:
Yes, I have finished the modification. Thanks!